Jetpack E2E: update expectations for `settings__media.ts`.

worldomonation opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/80453.

## Proposed Changes

This PR updates the expectations of `test/e2e/specs/media/settings__media.ts` to pass in all scenarios:

| environment  | jetpack-deployment |  jetpack-production   | 
|--------------|--------------|--------------|
| Simple       | ✔️      | ✔️      | 
| AT       | ✔️    | ✔️      |

Key changes:
- remove jetpack-remote-site codepath because the Jetpack test model has changed to focus on Dotcom-hosted sites.
- add description and comments to the test suite, along with keywords.
- guard the "Toggle Carousel" step behind a AT check, as this option is not present on AT sites. **It is uncertain whether this is expected.**

See the screenshot below - left is an AT site, right is Simple.
![image](https://github.com/Automattic/wp-calypso/assets/6549265/1a776da7-5f56-498f-a517-b1b141b63f25)

## Testing Instructions

If possible, run the `Jetpack Atomic Deployment E2E Tests (desktop)` build on the branch.

When run locally:

AT, jetpack=wpcom-deployment
```bash
➜  ~/workspace/wp-calypso-second git:(fix/e2e-editor-page-basic-flow-80543) ✗ TEST_ON_ATOMIC=true ATOMIC_VARIATION=default RUN_ID='Atomic: default' JETPACK_TARGET=wpcom-deployment yarn workspace wp-e2e-tests test -- test/e2e/specs/media/settings__media.ts
 PASS  specs/media/settings__media.ts (17.097 s)
  Jetpack Settings: Media (Atomic: default)
    ✓ Navigate to Settings > Media (10467 ms)
    ✓ Toggle photo metadata (114 ms)
    ✓ Select "Black" for carousel background color (19 ms)
    ○ skipped Toggle Carousel

Test Suites: 1 passed, 1 total
Tests:       1 skipped, 3 passed, 4 total
Snapshots:   0 total
Time:        17.265 s, estimated 30 s
```

AT, jetpack=wpcom-production
```
➜  ~/workspace/wp-calypso-second git:(fix/jetpack-atomic-settings-media) ✗ TEST_ON_ATOMIC=true ATOMIC_VARIATION=default RUN_ID='Atomic: default' JETPACK_TARGET=wpcom-production yarn workspace wp-e2e-tests test -- test/e2e/specs/media/settings__media.ts
 PASS  specs/media/settings__media.ts (14.247 s)
  Jetpack Settings: Media (Atomic: default)
    ✓ Navigate to Settings > Media (7175 ms)
    ✓ Toggle photo metadata (86 ms)
    ✓ Select "Black" for carousel background color (22 ms)
    ○ skipped Toggle Carousel

Test Suites: 1 passed, 1 total
Tests:       1 skipped, 3 passed, 4 total
Snapshots:   0 total
Time:        14.415 s
```

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

worldomonation committed:
test/e2e/specs/media/settings__media.ts
- remove jetpack-remote-site codepath.
- add description and comments.
- guard the "Toggle Carousel" step behind a AT check, as this option is not present on AT sites.

worldomonation commented on test/e2e/specs/media/settings__media.ts:
```diff
@@ -32,56 +37,25 @@ describe( DataHelper.createSuiteTitle( 'Jetpack Settings: Media' ), function ()
 		await testAccount.authenticate( page );
 	} );
 
-	if ( envVariables.JETPACK_TARGET === 'remote-site' ) {
```
My understanding is that we're focusing on the Jetpack experience as hosted on dotcom, which is why I felt comfortable removing this clause.

dpasque replied:
Eh, my instinct would be to actually leave this in for now. It does complicate the test a little bit, but the remote, connected site case is something we might want to target in the future, and we have all the support elsewhere for! But if you feel strongly that you want to take it out, that's okay too 🙂

worldomonation replied:
I've restored the `remote-site` portion.

worldomonation committed:
packages/calypso-e2e/src/data-helper.ts
- add `getWPAdminURL` method to assemble url.

test/e2e/specs/media/settings__media.ts
- revise to wait for `wp-admin` for the ecomm-plan user.
- work around WPAdmin home page.

worldomonation commented:
The changes in this PR work, as shown by the `settings__media.ts` not showing up in the list of failed tests.

![image](https://github.com/Automattic/wp-calypso/assets/6549265/cf04355c-e24c-4847-8e5c-c4f62c5513b9)

worldomonation committed:
packages/calypso-e2e/src/data-helper.ts
- change default behavior of `getAccountSiteURL` to not return a protocol by default.

packages/calypso-e2e/src/lib/test-account.ts
- similar to above, change default behavior of `getSiteURL` to not return protocol.

test/e2e/specs/media/settings__media.ts
- replace direct access to primary test site slug with `testAccount.getSiteURL` method.

worldomonation committed:
packages/calypso-e2e/src/test/data-helper.test.ts
- update unit test expectations

worldomonation committed:
packages/calypso-e2e/src/data-helper.ts
- handle more edge cases.

packages/calypso-e2e/src/test/data-helper.test.ts
- add tests.

dpasque commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -172,7 +190,7 @@ export function getAccountCredential( accountType: TestAccountName ): AccountCre
 export function getAccountSiteURL(
 	accountType: TestAccountName,
-	{ protocol = true }: { protocol?: boolean } = {}
+	{ protocol = false }: { protocol?: boolean } = {}
```
**Fix (blocking)**: As a heads up, it looks like this change in the default might break the P2 spec here:
https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/specs/p2/p2__post.ts#L32

... Which relies on the protocol being added by default. I think it's fine either way the default goes, just wanted to avoid the breaking change!

worldomonation replied:
Good catch, thanks. I'll take a look.

worldomonation replied:
Change removed from the PR.

dpasque commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -139,6 +139,24 @@ export function getCalypsoURL(
+ * @returns {string} The full URL path.
+ * @throws {Error} If siteSlug is empty.
+ */
+export function getWPAdminURL( siteSlug: string, path: string ): string {
+	if ( ! siteSlug ) {
+		throw new Error( 'siteSlug must not be empty.' );
+	}
+	if ( siteSlug.match( /^(http:\/\/|https:\/\/)/ ) ) {
+		siteSlug = siteSlug.replace( /^(http:\/\/|https:\/\/)/, '' );
+	}
+	return new URL( path, `https://${ siteSlug }` ).href;
+}
+
```
**Question/suggestion (non-blocking)**: Could you explain the purpose of this function a little bit more?

I think after reading the name/signature, I was expecting it to do a bit more than it does, which is really just safely join a path and hostname. I think that is something that could be safely offloaded to the caller as they would know if they are using a protocol already, and could just join the path the site host name themselves.

I wonder if a better abstraction would be to move this function to the `TestAccount` class? And then it could handle fetching the site name for you, and always set `protocol: true`.

And since it's called `getWPAdminURL` -- should the user have to specify the `wp-admin` path themselves? Or would it be better to have the function handle creating that part of the path? Otherwise we could just rename this to something like `getSiteUrl` since it could then be used for both `wp-admin` paths and non-admin paths!

worldomonation replied:
The intention of `getWPAdminURL` is to be a counterpart to the `getCalypsoURL` that we use almost everywhere. 

> I think that is something that could be safely offloaded to the caller as they would know if they are using a protocol already, and could just join the path the site host name themselves.

I'm a little less trusting of the caller. The way I see it is that if it's up to the caller how to assemble the WP-Admin URL, we will see a bunch of disparate ways it gets implemented. Not exact analogue, but we already see this with a Page POM's `visit` method; there are a few different signatures for the method[1].

> I wonder if a better abstraction would be to move this function to the TestAccount class? And then it could handle fetching the site name for you, and always set protocol: true.

There's precedent here with `TestAccount.getSiteURL` which in turn calls `DataHelper.getAccountSiteURL`. I am tepid on moving the `getCalypsoURL` and `getWPAdminURL` methods into the test account class though, because what if we want to construct URLs without being tied to a specific account? Having it in the DataHelper class keeps the function less coupled to the test accounts themselves.

[1] Examples:
- https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts#L97
- https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/pages/general-settings-page.ts#L41C52-L41C52
- https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/pages/plugins-page.ts#L81

worldomonation replied:
I've removed the changes relating to `getWPAdminURL` from this PR, because this risks ballooning in size. It's also not something absolutely required to fix the Jetpack tests for now.

dpasque reviewed (approved):
Thanks for fixing this! 😄  I left a few comments, and as a heads-up, one of them is blocking (the default protocol value change may break our P2 tests), but this shouldn't need to be re-reviewed, so dropping an approval. 👍

worldomonation committed:
test/e2e/specs/media/settings__media.ts
- restore previously deleted portion of the spec for remote-sites.

packages/calypso-e2e/src/data-helper
- remove addition of getWPAdminURL method.

packages/calypso-e2e/src/lib/test-account.ts
- add more descriptive JSDoc

worldomonation committed:
packages/calypso-e2e/src/lib/test-account.ts
- fix JSDoc
